bond_app/oauth2_state_store.py

```diff
@@ -0,0 +1,63 @@
+import base64
+import json
+import secrets
+from collections import namedtuple
+from google.cloud import ndb
+
+# Information associated with a tokens for refreshing service account credentials.
+OAuth2StateInfo = namedtuple('OAuth2StateInfo', ['nonce'])
+
+
+class OAuth2State(ndb.Model):
+    """
+    Model used to store entries in Datastore.
+
+    This is not used as the return type for OAuth2StateStore because it depends on ndb, which we want to stub out for unit
+    tests.
+    """
+    nonce = ndb.StringProperty()
+
+    @classmethod
+    def kind_name(cls):
+        return cls.__name__
+
+
+class OAuth2StateStore:
+    """
+    Stores OAuth2 State nonces for csrf protection.
+    """
+
+    def save(self, user_id, provider, nonce):
+        """
+        Persists a RefreshToken by creating a new entity or updating an existing entity with the same id
+        :param user_id
+        :param provider:
+        :param nonce: random value for csrf protection
+        """
+        ouath2_nonce = OAuth2State(key=OAuth2StateStore._oauth2_state_store_key(user_id, provider), nonce=nonce)
+        ouath2_nonce.put()
```

Comment:
what does this do?

Reply:
oh also typo heh

Reply:
oh i see this persists it in the datastore, what is the underlying db we are using here?

Reply:
There is no underlying _real_ DB. We're using https://cloud.google.com/datastore/

Reply:
ok gotcha 👍